python/jupyterlab-chat/jupyterlab_chat/ychat.py

```diff
@@ -52,105 +55,119 @@ def yusers(self) -> Map:
     def ymetadata(self) -> Map:
         return self._ymetadata
 
-    def get_user(self, username: str) -> Optional[dict[str, str]]:
+    def get_user(self, username: str) -> Optional[User]:
         """
-        Returns a message from its id, or None
+        Returns a user from its id, or None
         """
-        return self.get_users().get(username, None)
+        user = self._get_users().get(username, None)
+        if user is None:
+            return None
+        else:
+            return User(**user)
 
-    def get_user_by_name(self, name: str) -> Optional[dict[str, str]]:
+    def get_user_by_name(self, name: str) -> Optional[User]:
         """
         Returns a user from its name property, or None.
         """
-        return next(
-            (user for user in self.get_users().values() if user["name"] == name),
+        user = next(
+            (user for user in self._get_users().values() if user["name"] == name),
             None
         )
+        if user is None:
+            return None
+        else:
+            return User(**user)
 
-    def get_users(self) -> dict[str, dict[str, str]]:
+    def _get_users(self) -> dict[str, dict]:
         """
         Returns the users of the document.
         :return: Document's users.
         """
         return self._yusers.to_py() or {}
 
-    def set_user(self, user: dict[str, str]) -> None:
+    def set_user(self, user: User) -> None:
         """
         Adds or modifies a user.
         """
         with self._ydoc.transaction():
-            self._yusers.update({user["username"]: user})
+            self._yusers.update({
+                user.username: asdict(user)
+            })
 
-    def get_message(self, id: str) -> tuple[Optional[dict], Optional[int]]:
+    def get_message(self, id: str) -> tuple[Optional[Message], Optional[int]]:
         """
         Returns a message and its index from its id, or None
         """
-        return next(
-            ((msg, i) for i, msg in enumerate(self.get_messages()) if msg["id"] == id),
+        message = next(
+            ((msg, i) for i, msg in enumerate(self._get_messages()) if msg["id"] == id),
             (None, None)
         )
+        if message[0] is None:
+            return (None, None)
+        else:
+            return (Message(**message[0]), message[1])
 
-    def get_messages(self) -> list[dict]:
+    def _get_messages(self) -> list[dict]:
         """
         Returns the messages of the document.
         :return: Document's messages.
         """
         return self._ymessages.to_py() or []
```

Comment:
> In this PR I changed YChat.get_messages() to "private", because it is not returning the list of Message but a list of dict, to avoid confusion.

Instead of making this method private, could we return a list of `Message` objects? This method is necessary to https://github.com/jupyterlab/jupyter-ai/pull/1151.

```suggestion
    def get_messages(self) -> list[Message]:
        """
        Returns the messages of the document.
        :return: Document's messages.
        """
        message_dicts = self._ymessages.to_py() or []
        return [Message(**message_dict) for message_dict in message_dicts]
```

The performance overhead of this change should be minor compared to the call to `self._ymessages.to_py()`.

Reply:
Other calls to `get_messages()` can also be updated accordingly, and will look much cleaner.

Reply:
Not sure about that, this method will be often used internally without the need of casting every message.
Maybe we could keep this one for internal use, and add a new public method that returns slice of messages as proper `Message` objects.
```python
get_messages(start=0, end=-1, step=1):
```

Reply:
I added a public method to get all the messages as a list of Messages.